pkg/agent/knapsack/queued_querier.go

```diff
@@ -0,0 +1,160 @@
+package knapsack
+
+import (
+	"container/list"
+	"context"
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/kolide/launcher/pkg/backoff"
+)
+
+type synchronousQuerier interface {
+	Query(query string) ([]map[string]string, error)
+}
+
+type queuedQuerier struct {
+	logger      log.Logger
+	cancel      context.CancelFunc
+	queueCh     chan struct{}
+	queue       *queue
+	syncQuerier synchronousQuerier
+}
+
+type queue struct {
+	items *list.List
+	mutex sync.Mutex
+}
+
+func (q *queue) push(item *queueItem) {
+	q.mutex.Lock()
+	defer q.mutex.Unlock()
+
+	q.items.PushBack(item)
+}
+
+func (q *queue) pop() *queueItem {
+	q.mutex.Lock()
+	defer q.mutex.Unlock()
+
+	e := q.items.Front() // First element
+	if e == nil {
+		return nil
+	}
+
+	q.items.Remove(e) // Dequeue
+
+	if e == nil {
+		return nil
+	}
+
+	item, ok := e.Value.(*queueItem)
+	if ok {
+		return item
+	}
+
+	return nil
+}
+
+type queueItem struct {
+	query    string
+	callback func(result []map[string]string, err error)
+}
+
+func NewQueuedQuerier(logger log.Logger) *queuedQuerier {
+	q := &queue{
+		items: list.New(),
+	}
+	qq := &queuedQuerier{
+		logger:  log.With(logger, "component", "querier"),
+		queue:   q,
+		queueCh: make(chan struct{}, 10),
```

Comment:
I arbitrarily chose 10, but I don't have a good feel for what's appropriate.

This would be the number of queries in the queue when `Query` invocations become blocking, and waiting for other queries to process. Maybe it should be larger just to provide plenty of room?

Reply:
hmm ... I worry there is a chance that some other part of the code gets blocked why waiting on the channel to drain. Why don't we just leave it unbuffered and put the channel send in the `Query` func in a goroutine?

```go
func (qq *queuedQuerier) Query(query string, callback func(result []map[string]string, err error)) error {
	if qq == nil {
		return errors.New("queuedQuerier is nil")
	}

	// Push the item on to the queue, and notify the channel so the query is processed
	qq.push(&queueItem{query: query, callback: callback})

	go func() {
		qq.queueCh <- struct{}{}
	}()

	return nil
}
```

Or we could set the buffer 1 and check that `len(qq.queueCh) == 0` before pusing

Reply:
Reasonable! This should only really get used for the internal stuff, so order of 10 seems okay. Comment please?

Reply:
Hmm I like @James-Pickett suggestion, it would be nice to guarantee that `Query` does not block.

Reply:
I think James is probably correct. There's already the list there.

It makes me wonder if this channel is needed.

Reply:
The channel is mostly to wake up the processing thread in cases where the queue is empty, or the querier is not yet available.

We could just spawn a goroutine for every query instead of queueing and processing serially... That would assume there is proper thread safety within the querier impls. If we spawn 100 goroutines all submitting queries, would that just work? That would also mean clients should not expect their queries to be run in order necessarily.

Reply:
> We could just spawn a goroutine for every query instead of queueing and processing serially... That would assume there is proper thread safety within the querier impls. If we spawn 100 goroutines all submitting queries, would that just work? That would also mean clients should not expect their queries to be run in order necessarily.

I was thinking the same ... effectively have a stack of blocking goroutines as an unordered queue. I don't think we care about order at this level.

If the caller wanted, they could force their own queries to by synchronous at their level. ie: send query A ... wait for results .. send query B